src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorShape.cs

```diff
@@ -273,7 +273,7 @@ ref Unsafe.As<nint, int>(ref metadata[rank * 2]),
 
         private TensorShape(nint flattenedLength, nint linearLength, scoped ReadOnlySpan<nint> lengths, scoped ReadOnlySpan<nint> strides, scoped ReadOnlySpan<int> linearRankOrder, int rank)
         {
-            Debug.Assert((linearLength == 0) || ((flattenedLength % linearLength) == 0));
```

Comment:
This one should be correct. If we're calling this constructor its because we've already validated everything and so either `flattenedLength == linearLength` (no broadcast) or `flattenedLength % linearLength == 0` (broadcast, so linearLength should be a perfect multiple of flattenedLength)

Reply:
In the case it is a broadcast, then it is very likely `linearLength <= flattenedLength`.

However, this might be missing the case of large strides, in which case `linearLength >= flattenedLength`. That would require us also taking in the flags and handling the scenario

Reply:
Its also called during TensorShape.Slice, in which case the linear length of the original memory may be longer than the new flattened length.

the check how it was before fails some valid slice checks. So maybe the new one isn't quite right, but the old one is off as well.

Reply:
As part of `Slice`, the `linearLength` should be updated to be the new `minimumLinearLength` so that this isn't an issue.

In general we don't expect `linearLength` to be the original backing storage length, as that is potentially dangerous and can allow "leaking" of details that weren't intended.

Instead, we expect it to track the amount of linear memory that is actually needed by the array. Thus an `int[]` of `length 20` can be used to create a `2x2` matrix and `linearLength` would be at least `1`, most likely `4`, and up to `20` with `flattenedLength` always being `4`.